### DISCUSSION TITLE
EmguCV Image Class issues

### DISCUSSION THREAD

**Discussion by onslauth** [emgucv/emgucv#749] [Q&A]
Hi,

**The first question:**
I am trying to create a an image with the following dimensions  "200x1" ( 1 pixel wide, 200 pixels high ). However, EmguCV seems to be creating an extra 3 columns in the image.

The code to demonstrate the issue:
```
Emgu.CV.Image< Hsv, Byte > test_img = new Emgu.CV.Image< Hsv, Byte >( 1, 200 );
test_img[ 0, 0 ] = new Hsv( 10, 10, 10 );
test_img[ 1, 0 ] = new Hsv( 20, 20, 20 );
```

Examining the structure, I see the following in the debugger:
![image](https://user-images.githubusercontent.com/44875/184835945-a146094c-457b-4f73-970a-ce3283b1aea9.png)

And looking at the data:
![image](https://user-images.githubusercontent.com/44875/184836125-6f3001bc-de7f-44bf-a5d4-d5822acff789.png)

Why are there 3 pixel spaces between the first entry and second entry?

**Second question:**
When converting from Mat to Image class using the following:
```
Image< Hsv, Byte > img = image_in_hsv_format.ToImage< Hsv, Byte >( );
img.Save( @"C:\test.jpg" );
```
The saved image is not the HSV format, but the BGR image. If I save the `image_in_hsv_format` using `CvInvoke.Imwrite`, the image comes out in HSV format. Is there a way to save the Image class in HSV format without it converting to BGR before saving?

**Comment by sn4k3** [comment]
> Why are there 3 pixel spaces between the first entry and second entry?

1 byte for each raw color.
HSV, RGB and BGR uses 3 bytes per pixel
So for RGB:
```
index 0 == pixel 1 Red
index 1 == pixel 1 Green
index 2 == pixel 1 Blue
index 3 == pixel 2 Red
...
```

RGBA and BGRA uses 4 bytes per pixel, but in some other libraries they use `uint` instead making 1 uint per pixel instead of 4 bytes per pixel

> When converting from Mat to Image class:

Why do you want to convert Mat to Image? Mat can also use `.Save()` method

**Reply by onslauth** [reply]
> 1 byte for each raw color. HSV, RGB and BGR uses 3 bytes per pixel So for RGB:

I understand this, but if you look at the code and screenshots provided, the first and second pixel are set as follows:
```
img[ 0, 0 ] = new Hsv( 10, 10, 10 );
img[ 1, 0 ] = new Hsv( 20, 20, 20 );
```
And the img is a single pixel wide, 200 pixel high Image.
![184836125-6f3001bc-de7f-44bf-a5d4-d5822acff789](https://user-images.githubusercontent.com/44875/187887798-a6da482a-467a-437c-a48e-4a3b8abe8155.png)
What are these bytes?

> Why do you want to convert Mat to Image?

I would like to access individual pixels in the Mat, and apparently this is not possible, or I have not found any documentation that describes how this can be done.

**Reply by sn4k3** [reply]
> What are these bytes?

I don't know, maybe any extra data required by HSV? Wait for author to awnser that.

> I would like to access individual pixels in the Mat, and apparently this is not possible

If so Mat would be a bit useless in many ways.
A quick search on google gives you multiple ways to do that, but it's no efficient. 
Good news is that Mat gives you the `IntPtr` of the data and with NET new memory direct access (`Span`) you can access those pixels in blazing fast without almost no cost! Prior to `Span` you could use unsafe `Pointers`.
Look at my extensions: https://github.com/sn4k3/UVtools/blob/master/UVtools.Core/Extensions/EmguExtensions.cs look that [GetDataByteSpan](https://github.com/sn4k3/UVtools/blob/master/UVtools.Core/Extensions/EmguExtensions.cs#L198) with that you can access pixels directly. If you want `[,]` format use Span2D(require dependency) with width and height properties.
So this beat Image performance all day!

My project is all built around `Mat`, give it a try, load a file with 2000 Mats and see how fast it is to process 2000 mats ;)

![UVtools_2022-09-01_18-16-48](https://user-images.githubusercontent.com/113281/187973996-04bb0a40-16b1-45b3-b217-79aa1cd1b8e0.png)

13s to fill 2150 Mat's that need to be decoded first and then all pixel iterated to set noise based on a pattern. That's fast!
[Operation](https://github.com/sn4k3/UVtools/blob/master/UVtools.Core/Operations/OperationPixelArithmetic.cs#L771)

**Comment by emgucv** [comment]
Image<,> class are based on the IplImage object. IplImage's width steps is always a multiple of 4 (bytes) for performance reason.

See this:
https://answers.opencv.org/question/20407/matstep-iplimagewidthstep/
